### PR TITLE
added support for finetuning GPT-3.5 and HFModel

### PR DESCRIPTION
**Pull request by arnavsinghvi11**
-support for finetuning OpenAI models with latest API
-modified BootstrapFinetune to support OpenAI fine-tuning
-modified BootstrapFinetune to support HFModel fine-tuning with HFClientTGI checkpoint integrations as well
-added model initialization validation for HFClientTGI within latest changes

### CODE REVIEW

File: dsp/modules/hf_client.py [hunk 1]
```diff
@@ -28,11 +28,12 @@
...
 class HFClientTGI(HFModel):
-    def __init__(self, model, port, url="http://future-hgx-1", **kwargs):
-        super().__init__(model=model, is_client=True)
+    def __init__(self, model, checkpoint, port, url="http://future-hgx-1", **kwargs):
+        super().__init__(model=model, checkpoint=checkpoint, is_client=True)
+        self.model = model
         self.url = f"{url}:{port}"
         self.headers = {"Content-Type": "application/json"}
-
+        self._verify_model_with_server()
```
Comment (okhat) [inline dsp/modules/hf_client.py:36]: is this even implemented?

File: dspy/teleprompt/finetune.py [hunk 2]
```diff
@@ -46,10 +47,14 @@
...
 class BootstrapFinetune(Teleprompter):
-    def __init__(self, metric=None, teacher_settings={}, multitask=True):
+    def __init__(self, metric=None, provider=None, teacher_settings={}, multitask=True, return_hf_model=True):
         self.metric = metric
+        self.provider = provider
         self.teacher_settings = teacher_settings
         self.multitask = multitask
+        self.return_hf_model = return_hf_model
+
+        assert (not self.multitask) == self.return_hf_model
```
Comment (sam-h-bean) [inline dspy/teleprompt/finetune.py:57]: The default arguments cause this assertion to fail? Seems undesirable

Comment (arnavsinghvi11) [reply dspy/teleprompt/finetune.py:57]: Thanks for the correction!

File: dspy/teleprompt/finetune.py [hunk 3]
```diff
@@ -47,7 +47,7 @@
 
 
 class BootstrapFinetune(Teleprompter):
-    def __init__(self, metric=None, provider=None, teacher_settings={}, multitask=True, return_hf_model=False):
+    def __init__(self, metric=None, provider=None, teacher_settings={}, multitask=True, return_hf_model=True):
```
Comment (sam-h-bean) [inline dspy/teleprompt/finetune.py:50]: This was a broken merge on my part I think. My mistake here.

**Pull request closed without merging**